src/services/search.service.ts

```diff
@@ -26,9 +27,17 @@ export class SearchService implements SearchServiceAbstraction {
     }
 
     clearIndex(): void {
+        this.indexedEntityId = null;
         this.index = null;
     }
 
+    async awaitAndClearIndex(): Promise<void> {
+        if (this.indexing && this.indexingPromise != null) {
+            await this.indexingPromise;
```

Comment:
help me understand the scenario and perhaps it's my limited understanding of async await in the TypeScript world, but if multiple call stacks are awaiting on a promise, or when that scenario in this instance would actually occur in the Angular/SPA or Desktop/nodeJS realm. I guess I'm not understanding when this scenario would actually occur and why then we'd be calling this again before it completed the first time (is indexing done in a background thread, etc.?) as it seems the prior implementation still `await`'ed the result.

Reply:
I wasn't able to get it to happen in a debugger, but I think what's going on is navigation while indexing results in triggering more indexing.

This is the flow that this is supposed to prevent:
We are indexing an organization, we navigate to a personal vault before indexing finishes. We'll clear out the index, and start indexing again. Ciphers indexing has a quick return in we're already indexing, so the second call will be ignored and the first index will complete. Only organization items will be indexed.

This method allows us to wait until indexing is done, clear it, and start a new job.

The other option is to rewrite the indexing method to be abortable and make clear index trigger an abort, but I couldn't think of a clean way to do that -- I don't think it's supported natively by typescript es5.

I hope that's clear. It's a little fuzzy in my head too, so I'm afraid I'm not explaining well

Reply:
That's a great explanation and now it makes sense, pesky navigation in the JavaScript web world, ruining lives since 1999.

🤔 , thinking through this scenario, can you instead build indexes to a specified `key`, so a hashtable of index references which get used based on the `key` of that index you wish to search on? I'm thinking in terms of search domains, where `lunr` has maybe 5 indexes built (each one on-demand the first time/lazy-loaded), one is your personal, and then perhaps 4 other orgs' vaults. So when searching the appropriate `key` is passed in, referencing the appropriate index.

Or, perhaps build `all the things` in one index, but add an index facet or filter based on the search scope (like an org id)?

Just thinking of ways we can avoid the invalid index state or aborting an index build due to navigation, etc.